File: Coding_Questions/C++/TwoNumberSum.cpp
//
// Created by John Perry on 11/29/20.
//

#include <vector>
#include <map>

// Solution 1: O(n^2) Time | O(1) Space
//std::vector<int> two_number_sum(std::vector<int> array, int target_sum) {
//    for (int i = 0; i < array.size(); ++i) {
//        for (int c = 0; c < array.size(); ++c) {
//            if (i == c) {
//                continue;
//            } else if (array[i] + array[c] == target_sum) {
//                return std::vector<int> {array[i], array[c]};
//            }
//        }
//    }
//    return {};
//}


// Solution 2: O(n) Time | O(n) Space
//std::vector<int> two_number_sum(std::vector<int> array, int target_sum) {
//    std::map<int, bool> nums;
//    for (const auto num : array) {
//        int possible_match = target_sum - num;
//        if (nums.count(possible_match) == 1) {
//            return std::vector<int> {num, possible_match};
//        } else {
//            nums[num] = true;
//        }
//    }
//    return {};
//}


// Solution 3: O(nlog(n)) Time | O(1) Space
std::vector<int> two_number_sum(std::vector<int> array, int target_sum) {
    std::sort(array.begin(), array.end());
    int left = 0;
    int right = array.size() - 1;
    while (left < right) {
        int possible_match = array[left] + array[right];
        if (possible_match == target_sum) {
            return std::vector<int> {array[left], array[right]};
        } else if (possible_match < target_sum) {
            ++left;
        } else if (possible_match > target_sum) {
            --right;
        }
    }
    return {};
}

File: Coding_Questions/Python/TwoNumberSum.py
from typing import List


# ---Two Number Sum---

# Solution 1: O(n^2) Time | O(1) Space
# def two_number_sum(array: List[int], target_sum: int) -> List[int]:
#     for main_index, i in enumerate(array):
#         for sub_index, c in enumerate(array):
#             if main_index == sub_index:
#                 continue
#             elif i + c == target_sum:
#                 return [i, c]
#     return list()


# Solution 2: O(n) Time | O(n) Space
# def two_number_sum(array: List[int], target_sum: int) -> List[int]:
#     nums = dict()
#     for num in array:
#         potential_match = target_sum - num
#         if potential_match in nums:
#             return [potential_match, num]
#         else:
#             nums[num] = True
#     return list()


# Solution 3: O(nlog(n)) Time | O(1) Space
def two_number_sum(array: List[int], target_sum: int) -> List[int]:
    array.sort()
    left = 0
    right = len(array) - 1

    while left < right:
        current_sum = array[left] + array[right]
        if current_sum == target_sum:
            return [array[left], array[right]]
        elif current_sum < target_sum:
            left += 1
        elif current_sum > target_sum:
            right -= 1
    return list()


print(two_number_sum([3, 5, -4, 8, 11, 1, -1, 6], 10))
# non-specific ordering => [-1, 11]
